File: src/main/java/dev/vxrp/bot/events/ModalListener.java
package dev.vxrp.bot.events;

import dev.vxrp.bot.events.modals.NoticeOfDeparture;
import dev.vxrp.bot.events.modals.Regulars;
import dev.vxrp.bot.events.modals.Support;
import dev.vxrp.bot.events.modals.Unban;
import net.dv8tion.jda.api.events.interaction.ModalInteractionEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.sql.SQLException;

public class ModalListener extends ListenerAdapter {
    public final Logger logger = LoggerFactory.getLogger(ModalListener.class);

    @Override
    public void onModalInteraction(ModalInteractionEvent event) {
        // Support
        if (event.getModalId().equals("supportTicket")) {
            Support.createSupportTicket(event, logger);
        }
        if (event.getModalId().equals("unbanTicket")) {
            Unban.createUnbanTicket(event, logger);
        }
        if (event.getModalId().startsWith("reason_action_unban_accept")) {
            try {
                Unban.acceptUnban(event);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
        if (event.getModalId().startsWith("reason_action_reason_dismiss")) {
            event.getJDA().retrieveUserById(event.getModalId().split(":")[1]).queue(user -> Unban.dismissUnban(event, user));
        }
        // Regulars
        if (event.getModalId().startsWith("regulars_data_modal")) {
            try {
                Regulars.regularDataModal(event);
            } catch (SQLException | InterruptedException e) {
                throw new RuntimeException(e);
            }
        }
        // Notice of Departure
        if (event.getModalId().equals("notice_of_departure")) {
            NoticeOfDeparture.createNewNoticeOfDeparture(event);
        }
        if (event.getModalId().startsWith("reason_action_accepted_nod")) {
            event.getJDA().retrieveUserById(event.getModalId().split(":")[1]).queue(user -> NoticeOfDeparture.acceptNoticeOfDeparture(event, user));
        }
        if (event.getModalId().startsWith("reason_action_dismiss_nod")) {
            event.getJDA().retrieveUserById(event.getModalId().split(":")[1]).queue(user -> {
                try {
                    NoticeOfDeparture.dismissNoticeOfDeparture(event, user);
                } catch (InterruptedException e) {
                    throw new RuntimeException(e);
                }
            });
        }
        if (event.getModalId().startsWith("reason_action_revoke_nod")) {
            event.getJDA().retrieveUserById(event.getModalId().split(":")[1]).queue(user -> {
                try {
                    NoticeOfDeparture.revokeNoticeOfDeparture(event, user);
                } catch (SQLException | InterruptedException e) {
                    throw new RuntimeException(e);
                }
            });
        }
    }
}

File: src/main/kotlin/dev/vxrp/bot/events/ModalListener.kt
/*
 * Copyright (c) 2024 Vxrpenter and the SCPToolsBot Contributors
 *
 * Licenced under the MIT License, any non-license compliant usage of this file(s) content
 * is prohibited. If you did not receive a copy of the license with this file, you
 *  may obtain the license at
 *
 *  https://mit-license.org/
 *
 *  This software may be used commercially if the usage is license compliant. The software
 *  is provided without any sort of WARRANTY, and the authors cannot be held liable for
 *  any form of claim, damages or other liabilities.
 *
 *  Note: This is no legal advice, please read the license conditions
 */

package dev.vxrp.bot.events

import dev.minn.jda.ktx.events.listener
import dev.minn.jda.ktx.messages.reply_
import dev.vxrp.bot.events.modals.ApplicationModals
import dev.vxrp.bot.events.modals.NoticeOfDepartureModals
import dev.vxrp.bot.events.modals.TicketModals
import dev.vxrp.bot.permissions.PermissionManager
import dev.vxrp.bot.permissions.enums.StatusMessageType
import dev.vxrp.configuration.data.Config
import dev.vxrp.configuration.data.Translation
import dev.vxrp.util.launch.LaunchOptionManager
import dev.vxrp.util.launch.enums.LaunchOptionSectionType
import dev.vxrp.util.launch.enums.LaunchOptionType
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.events.interaction.ModalInteractionEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import org.slf4j.LoggerFactory

class ModalListener(val api: JDA, val config: Config, val translation: Translation) : ListenerAdapter() {
    private val logger = LoggerFactory.getLogger(ModalListener::class.java)

    init {
        api.listener<ModalInteractionEvent> { event ->
            val launchOptionManager = LaunchOptionManager(config, translation)

            if (launchOptionManager.checkSectionOption(LaunchOptionType.MODAL_LISTENER, LaunchOptionSectionType.TICKET_MODALS).engage && event.modalId.startsWith("ticket")) {
                PermissionManager(config, translation).checkStatus(StatusMessageType.PANEL, config.ticket.settings.ticketLogChannel != "")?.let { embed ->
                    event.reply_("", listOf(embed)).setEphemeral(true).queue()
                } ?: TicketModals(logger, event, config, translation).init()
            }

            if (launchOptionManager.checkSectionOption(LaunchOptionType.MODAL_LISTENER, LaunchOptionSectionType.APPLICATION_MODALS).engage && event.modalId.startsWith("application")) {
                PermissionManager(config, translation).checkStatus(StatusMessageType.PANEL, config.ticket.settings.applicationMessageChannel != "")?.let { embed ->
                    event.reply_("", listOf(embed)).setEphemeral(true).queue()
                } ?: ApplicationModals(event, config, translation)
            }

            if (launchOptionManager.checkSectionOption(LaunchOptionType.MODAL_LISTENER, LaunchOptionSectionType.NOTICE_OF_DEPARTURE_MODALS).engage && event.modalId.startsWith("notice_of_departure")) {
                PermissionManager(config, translation).checkStatus(StatusMessageType.PANEL, config.settings.noticeOfDeparture.active)?.let { embed ->
                    event.reply_("", listOf(embed)).setEphemeral(true).queue()
                } ?: NoticeOfDepartureModals(event, config, translation).init()
            }
        }
    }
}